docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+
+## Out of scope
+
+1. Routing different kinds of data to different clouds, e.g: all telemetry to azure and all commands from/to Cumulocity.
+   Even though this requirement is realistic, thin-edge MQTT topics must not be corrupted with cloud specific semantics,
+   and the same requirement will be handled with some external routing mechanism(e.g: routing via bridge topics)
+
+# Proposals
+
+## Proposal 1: Dedicated topics for devices and services
+
+This proposal is built on the following assumptions/constraints:
+* thin-edge can identify the child devices (the nested ones as well), and all the services with unique IDs.
+* All child-devices connected to a parent device will have unique IDs under its parent namespace,
+  but not necessarily across the entire tedge namespace/hierarchy.
+* If all child devices **can not** guarantee that uniqueness across the entire hierarchy,
+  then they are expected to register with thin-edge to get that unique ID generated.
+* Given that unique ID, thin-edge can identify its entire parent lineage and descendent child devices.
+  Thin-edge internally maintains this mapping information as child devices register with it.
+* If all child devices **can** guarantee ID uniqueness across the entire hierarchy,
+  they can skip this registration step.
+* Services need not have unique IDs across the entire hierarchy.
+  They just need to have unique IDs under their parent's namespace.
+
+The proposal is as follows:
+
+Topics to have device id as the top level prefix with distinction on the target: device or service as in:
+
+```
+tedge/<device-id>/<target-type>/...
+```
+
+The following demonstrates the different topic prefix for each of the target types:
+
+|Target|Topic Prefix|
+|------|-------|
+|tedge device|`tedge/main/device/`|
+|service on tedge device|`tedge/main/service/<service-id>/`|
+|child device|`tedge/<child-id>/device/`|
+|service on child device|`tedge/<child-id>/service/<service-id>/`|
+
+Where `main` is used as an alias for the tedge device id.
+
+**Why have the `/device/` subtopic level?**
+
+To make a clear distinction between `device` data and `service` data.
+This distinction makes it easier to subscribe to all device only data excluding the service data
+with simple subscription filters like `tedge/main/device/#`.
+Without that device suffix, something like `tedge/main/#` would include the services' data as well.
+
+
+**Why have the `/service/` subtopic level and not have the <service-id> directly?**
+
+Primarily for future-proofing as `service` is a new kind of child abstraction that we've added now.
+If another abstraction is introduced in the future, say `plugin`, they can be namespaced under a `plugin/` subtopic.
+It also makes the distinction from `device` data clearer.
+
+### Telemetry
+
+For telemetry data, the topics would be grouped under a `telemetry/` sub-topic as follows:
+
+|Type|Example|
+|------|-------|
+|Measurements|`<target_prefix>/telemetry/measurements`|
```

Comment:
We should add the type info to the measurements (as it was also mentioned as part of the MUST HAVES and for consistency with other topics).

```suggestion
|Measurements|`<target_prefix>/telemetry/measurements/<measurement-type>`|
```

Reply:
This has been adopted:

- `'te/device/main///m/'` for a measurement with no type
- `te/device/main///m/battery_reading` as an example with a type.